### PR TITLE
[SYCL] Add clang implementation for accessor property no_alias

### PR DESCRIPTION
**Pull request by schittir**
no_alias is a compile-time-constant accessor property.  
It exists in cl::sycl::ONEAPI::property scope. 
When applied the LLVM IR should contain the noalias parameter attribute
on the kernel argument corresponding to that accessor.
This patch uses clang's RestrictAttr to emit `noalias` i.e., llvm::Attribute::NoAlias.

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -1515,6 +1515,9 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
   // MDNode for the intel_buffer_location attribute.
   SmallVector<llvm::Metadata *, 8> argSYCLBufferLocationAttr;
 
+  // MDNode for accessor no_alias property
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CodeGenModule.cpp:1518]:
```suggestion
  // MDNode for accessor no_alias property.
```

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 2]
```diff
@@ -1622,17 +1625,25 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
                     SYCLBufferLocationAttr->getLocationID()))
               : llvm::ConstantAsMetadata::get(CGF->Builder.getInt32(-1)));
 
+      if (parm->hasAttr<RestrictAttr>())
+      argAccessorNoAliasPropertyAttr.push_back(llvm::ConstantAsMetadata::get(
+          CGF->Builder.getInt1(parm->hasAttr<RestrictAttr>())));
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CodeGenModule.cpp:1630]:
```suggestion
        argAccessorNoAliasPropertyAttr.push_back(llvm::ConstantAsMetadata::get(
            CGF->Builder.getInt1(true)));
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -1717,11 +1721,21 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
...
       QualType PropTy = Prop->getAsType();
+      if (Util::isSyclAccessorNoAliasPropertyType(PropTy)) 
+        handleNoAliasProperty(Param, PropTy, Loc);
       if (Util::isSyclBufferLocationType(PropTy))
         handleBufferLocationProperty(Param, PropTy, Loc);
     }
   }
 
+  void handleNoAliasProperty(ParmVarDecl *Param, QualType PropTy,
+                                    SourceLocation Loc) {
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1732]: Formatting looks off here.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -4245,6 +4259,20 @@ bool Util::isSyclKernelHandlerType(const QualType &Ty) {
   return matchQualifiedTypeName(Ty, Scopes);
 }
 
+bool Util::isSyclAccessorNoAliasPropertyType(const QualType &Ty) {
+  const StringRef &PropertyName = "no_alias";
+  const StringRef &InstanceName = "instance";
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:4264]:
```suggestion
bool Util::isSyclAccessorNoAliasPropertyType(QualType Ty) {
  StringRef PropertyName = "no_alias";
  StringRef InstanceName = "instance";
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -4245,6 +4259,20 @@ bool Util::isSyclKernelHandlerType(const QualType &Ty) {
...
+  std::array<DeclContextDesc, 6> Scopes = {
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:4267]:
```suggestion
      Util::DeclContextDesc{Decl::Kind::Namespace, "cl"},
      Util::DeclContextDesc{Decl::Kind::Namespace, "sycl"},
```

**Review by AaronBallman — COMMENTED** [review]
Can you also remove [WIP] from the title and add [SYCL] instead?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:4264]: @AaronBallman, why are these variables needed?  Can't the literals be used directly below?

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp]: I was sort of wondering that myself -- I think the variables can just go away.

**Comment by schittir** [comment]
I'm not sure why the build bot lit test runs fail. They all pass on my local workspace.

**Comment by AaronBallman** [comment]
> I'm not sure why the build bot lit test runs fail. They all pass on my local workspace.

You should try catching the PR up to upstream. The testing issue may have been resolved elsewhere, or it'll restart CI in case the issues are transient.

**Comment by schittir** [comment]
> > I'm not sure why the build bot lit test runs fail. They all pass on my local workspace.
> 
> You should try catching the PR up to upstream. The testing issue may have been resolved elsewhere, or it'll restart CI in case the issues are transient.

Caught up PR to upstream and resolved failing tests.

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp [hunk 6]
```diff
@@ -29,8 +29,7 @@ int main() {
...
 }
 
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel1"() #0 !kernel_arg_buffer_location ![[NUM4:[0-9]+]] !disable_loop_pipelining ![[NUM5:[0-9]+]]
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel2"() #0 !kernel_arg_buffer_location ![[NUM4]]
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel3"() #0 !kernel_arg_buffer_location ![[NUM4]] !disable_loop_pipelining ![[NUM5]]
-// CHECK: ![[NUM4]] = !{}
+// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel1"() #0 {{.*}} !disable_loop_pipelining ![[NUM5:[0-9]+]]
+// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel2"() #0 {{.*}}
+// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel3"() #0 {{.*}} !disable_loop_pipelining ![[NUM5]]
 // CHECK: ![[NUM5]] = !{i32 1}
```
Comment (smanna12) [inline clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:35]: This seems unrelated to the patch. Is there any reason you are modifying the lit test?

File: clang/test/CodeGenSYCL/initiation_interval.cpp [hunk 7]
```diff
@@ -39,11 +39,10 @@ int main() {
...
 
-// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name1"() #0 !kernel_arg_buffer_location ![[NUM0:[0-9]+]] !initiation_interval ![[NUM1:[0-9]+]]
-// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name2"() #0 !kernel_arg_buffer_location ![[NUM0]] !initiation_interval ![[NUM42:[0-9]+]]
-// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name3"() #0 !kernel_arg_buffer_location ![[NUM0]] !initiation_interval ![[NUM2:[0-9]+]]
-// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name4"() #0 !kernel_arg_buffer_location ![[NUM0]]
-// CHECK: ![[NUM0]] = !{}
+// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name1"() #0 {{.*}} !initiation_interval ![[NUM1:[0-9]+]]
+// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name2"() #0 {{.*}} !initiation_interval ![[NUM42:[0-9]+]]
+// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name3"() #0 {{.*}} !initiation_interval ![[NUM2:[0-9]+]]
+// CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name4"() #0 {{.*}}
```
Comment (smanna12) [inline clang/test/CodeGenSYCL/initiation_interval.cpp:45]: same for here.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 8]
```diff
@@ -1687,17 +1690,25 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
...
-  if (LangOpts.SYCLIsDevice && !IsEsimdFunction)
+  if (LangOpts.SYCLIsDevice && !IsEsimdFunction) {
+
+    Fn->setMetadata("noalias", llvm::MDNode::get(
+                                   VMContext, argAccessorNoAliasPropertyAttr));
+
     Fn->setMetadata("kernel_arg_buffer_location",
                     llvm::MDNode::get(VMContext, argSYCLBufferLocationAttr));
-  else {
+  } else {
```
Comment (MrSidims) [inline clang/lib/CodeGen/CodeGenModule.cpp:1711]: Question to @premanandrao and @DenisBakhvalov : does it meant to be, that kernel_arg_* metadata is being applied to ESIMD functions (aka shall `else` be replaces with `else if`?
Also, I think, that `noalias` function parameter attribute is quiet useful optimization hint and shall be emitted for ESIMD as well.

**Review by MrSidims — COMMENTED** [review]
Overall LGTM

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp (hunk 6 above)
Comment (schittir) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:35]: The position of !kernel_arg_buffer_location has changed because I inserted !noalias in the string, due to which this test is failing.

Comment (smanna12) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:35]: I see. Thanks for the explanation.

Comment (MrSidims) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:35]: It has to be modified anyway, since `noalias` metadata starts to be generated. The way the test is modified in this patch is IMO preferable, since disable_loop_pipelining.cpp shan't check neither buffer_location nor noalias metadata.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 8 above)
Comment (schittir) [reply clang/lib/CodeGen/CodeGenModule.cpp:1711]: Your suggestion makes sense. I am open to changing it to `else` ; it wasn't clear to me from the spec description about which way to go.

File: clang/test/CodeGenSYCL/initiation_interval.cpp (hunk 7 above)
Comment (schittir) [reply clang/test/CodeGenSYCL/initiation_interval.cpp:45]: Same response here.

**Review by MrSidims — CHANGES_REQUESTED** [review]
Wait a second. Wasn't no_alias property supposed to generate function parameter attribute, instead of a metadata?

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp (hunk 6 above)
Comment (smanna12) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:35]: I agree. Thanks @MrSidims for the explanation.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 8 above)
Comment (DenisBakhvalov) [reply clang/lib/CodeGen/CodeGenModule.cpp:1711]: > does it meant to be, that kernel_arg_* metadata is being applied to ESIMD functions

I believe it is still a requirement for ESIMD kernels.

**Comment by schittir** [comment]
> Wait a second. Wasn't no_alias property supposed to generate function parameter attribute, instead of a metadata?

Thanks for the catch. Changed it to generate a function parameter attribute instead of metadata.

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp [hunk 9]
```diff
@@ -29,8 +29,7 @@ int main() {
   return 0;
 }
 
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel1"() #0 !kernel_arg_buffer_location ![[NUM4:[0-9]+]] !disable_loop_pipelining ![[NUM5:[0-9]+]]
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel2"() #0 !kernel_arg_buffer_location ![[NUM4]]
-// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel3"() #0 !kernel_arg_buffer_location ![[NUM4]] !disable_loop_pipelining ![[NUM5]]
-// CHECK: ![[NUM4]] = !{}
+// CHECK: define dso_local spir_kernel void @"{{.*}}test_kernel1"() #0 {{.*}} !disable_loop_pipelining ![[NUM5:[0-9]+]]
```
Comment (MrSidims) [inline clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:32]: No need in the tests changes then (at least no in this patch)

**Review by MrSidims — COMMENTED** [review]
LGTM, but you shall change the description.

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp (hunk 9 above)
Comment (smanna12) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp:32]: I will fix this in a separate patch.

**Review by premanandrao — COMMENTED** [review]
LGTM, but please check the pre-commit run-time failure.

**Comment by smanna12** [comment]
LGTM. I looked at the pre-commit failure:

http://icl-jenkins.sc.intel.com:8080/blue/organizations/jenkins/SYCL_CI%2Fintel%2FLin%2FLLVM_Test_Suite/detail/LLVM_Test_Suite/3272/pipeline

Failed Tests (1):
 SYCL :: Reduction/reduction_nd_N_vars.cpp

@schittir, you can try restarting jenkins/precomit testing.

**Comment by bader** [comment]
> @schittir, you can try restarting jenkins/precommit testing.

This should not be the solution for pre-commit failures. The test must pass w/o re-starts. Please, investigate why the test gives wrong results.
Tagging @v-klochkov for awareness.

File: clang/test/CodeGenSYCL/disable_loop_pipelining.cpp (hunk 9 above)
Comment (schittir) [reply clang/test/CodeGenSYCL/disable_loop_pipelining.cpp]: Removed tests

**Comment by schittir** [comment]
> LGTM, but you shall change the description.

Done.

**Review by MrSidims — DISMISSED** [review]

File: clang/test/CodeGenSYCL/accessor_no_alias_property.cpp [hunk 10]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -emit-llvm %s -o - | FileCheck %s
+
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/accessor_no_alias_property.cpp:2]: Please add a comment here, indicating what this test tests for.

File: clang/test/CodeGenSYCL/accessor_no_alias_property.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+// CHECK: define {{.*}}spir_kernel void @_ZTSZ4mainE15kernel_function({{.*}} noalias {{.*}} %_arg_, {{.*}})
+
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/accessor_no_alias_property.cpp:4]: Could you also add a case where the accessor has some other property and verify that the kernel *does not* have the noalias attribute?

**Review by premanandrao — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]
LGTM

File: clang/lib/Sema/SemaSYCL.cpp [hunk 12]
```diff
@@ -4313,6 +4325,18 @@ bool Util::isSyclKernelHandlerType(const QualType &Ty) {
...
 
+bool Util::isSyclAccessorNoAliasPropertyType(QualType Ty) {
+  std::array<DeclContextDesc, 6> Scopes = {
+      Util::DeclContextDesc{Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "sycl"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "ONEAPI"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "property"},
+      Util::DeclContextDesc{Decl::Kind::CXXRecord, "no_alias"},
+      Util::DeclContextDesc{Decl::Kind::ClassTemplateSpecialization,
+                            "instance"}};
```
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp:4336]: I have to ask this to sleep better at night. 
Do we use this method because there is no way to write a partial specialization for the accessor class, which uses `nolias` attribute?

I think we are trying to avoid such runtime library declarations matchers and use more robust approach with a new attribute - https://github.com/intel/llvm/pull/2091. I would expect this `property` will be moved from `ONEAPI` to `sycl` namespace in the future and it would be great to avoid compiler changes for such library API updates.

Comment (schittir) [reply clang/lib/Sema/SemaSYCL.cpp:4336]: @bader The idea behind #2091 is not fully clear to me i.e., I am not sure how this idea could be applied to accessor properties. Could you please elaborate? 

At the moment, the way to understand each of the accessor properties(buffer_location @MrSidims, noalias etc.,) seems, in a straightforward way, to just check for its name and handle it based on that, since there is no indication in the spec regarding the change of scope. Are you suggesting adding an attribute to identify/indicate the accessor property parameter?

Comment (schittir) [reply clang/lib/Sema/SemaSYCL.cpp:4336]: @bader 
I understand the merit of avoiding the fragility caused by relying on SYCL header scope gathered from the DeclContext; This needs to be addressed for not only no_alias, but also buffer_location and other (future) accessor properties such as no_offset and noinit. This can be addressed using an attribute consistently for all the accessor properties. I can create a follow-up task/tracker to resolve this issue separately. Please let me know your thoughts. Thank you.

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:4438]: I agree with @bader that this is a fragile approach that ties the FE very tightly to the SYCL header files, but I don't think there's anything that can be done about it as part of this patch and so no changes are needed. Once we have the better way available to us, we should hit up all of these properties at once. @bader, please correct me if I'm wrong.

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/lib/Sema/SemaSYCL.cpp (hunk 12 above)
Comment (bader) [reply clang/lib/Sema/SemaSYCL.cpp:4438]: I agree, I just wanted to check if anyone has explored an option to support `no_alias` accessor property by specializing accessor class template to re-use existing `noalias` attribute.

Comment (schittir) [reply clang/lib/Sema/SemaSYCL.cpp:4438]: https://github.com/intel/llvm/issues/3529 - follow up issue

Comment (bader) [reply clang/lib/Sema/SemaSYCL.cpp:4438]: > Would you suggest adding an attribute to identify/indicate the accessor property parameter, maybe at a later time?

According to my understanding this patch applies `noalias` attribute to an accessor pointer member when accessor is instantiated with special template parameter. `noalias` attribute is already exposed to source level, so I'm wondering why can't we just apply it in DPC++ headers instead? This way, no compiler changes are required (maybe except propagating attribute to the generated kernel function parameters).
I guess someone from the runtime team should be able to answer this. @romanovvlad, do you know who can help to clarify this?

File: clang/test/CodeGenSYCL/accessor_no_alias_property.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -emit-llvm %s -o - | FileCheck %s
+// check that noalias parameter attribute is emitted when no_alias accessor property is used
+// CHECK: define {{.*}}spir_kernel void @_ZTSZ4mainE16kernel_function1({{.*}} noalias {{.*}} %_arg_, {{.*}})
+
+// check that noalias parameter attribute is NOT emitted when it is not used
+// CHECK: define {{.*}}spir_kernel void @_ZTSZ4mainE16kernel_function2{{.*}} !kernel_arg_buffer_location
+// CHECK-NOT: define {{.*}}spir_kernel void @_ZTSZ4mainE16kernel_function2({{.*}} noalias {{.*}}
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/accessor_no_alias_property.cpp:7]: Lines 6 and 7 here are broken.  The first checks to see if that kernel is defined, and the 2nd just checks to see if it is defined AGAIN.  This is not correct.  Presumably there needs to be a 'CHECK-NOT'/'CHECK-SAME' combo, like in the example here: 
https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-same-directive

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/accessor_no_alias_property.cpp:7]: PR - https://github.com/intel/llvm/pull/4283
